src/Microsoft.Identity.Web.TokenAcquisition/MicrosoftIdentityOptions.cs

```diff
@@ -93,8 +93,22 @@ public class MicrosoftIdentityOptions : OpenIdConnectOptions
         /// Performance improvements when working with MSAL only apps.
         /// Set to true if you have a shared cache with ADAL apps.
         /// </summary>
-        /// The default is <c>false.</c>
-        public bool LegacyCacheCompatibilityEnabled { get; set; }
+        /// <remarks>
+        /// The default is <c>false</c>.
+        /// </remarks>
+        public bool LegacyCacheCompatibilityEnabled { get; set; }
+
+        /// <summary>
+        /// When set to <c>true</c>, MSAL will lock cache access at the ConfidentialClientApplication level, i.e.
+        /// the block of code between BeforeAccessAsync and AfterAccessAsync callbacks will be synchronized. 
+        /// Apps can set this flag to <c>false</c> to enable an optimistic cache locking strategy, which may result in better performance
+        /// at the cost of cache consistency. 
+        /// Setting this flag to <c>false</c> is only recommended for apps which create a new ConfidentialClientApplication per request.
+        /// </summary>
+        /// <remarks>
+        /// The default is <c>false</c>.
+        /// </remarks>
+        public bool EnableCacheSynchronization { get; set; }
```

Comment:
There are cases, for example [here](https://github.com/AzureAD/microsoft-identity-web/blob/master/src/Microsoft.Identity.Web.TokenAcquisition/ServiceCollectionExtensions.cs#L40) where singletons are used. Can we force this setting to `true` there?

I think it's better to err on the side of correctness and not on the side of "perf" (btw, I don't think there is a sensible perf improvement that this makes, historically we created this setting to work around a deadlock). 

If it's too complex to keep track of object lifetime (i.e. some users might use CCA singleton via config), then it's better to set this to `true` as default.

Reply:
Thanks
MicrosoftIdentityOptions is not the recommended way in 2.x. we recommended customers to use MicrosoftIdentityApplicationOptions.

Is a boolean exposed in Msal?

Reply:
Here: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/1b56a35f5add14c9bb439d50b3ae95368f5c107d/src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationOptions.cs#L46